File: android/src/main/java/org/jellyfin/apiclient/interaction/AndroidDevice.java
package org.jellyfin.apiclient.interaction;

import android.content.Context;
import android.os.Build;
import android.provider.Settings;

import org.jellyfin.apiclient.interaction.device.IDevice;

public class AndroidDevice implements IDevice {
    private final Context context;
    private final String deviceId;
    private final String deviceName;

    public AndroidDevice(Context context) {
        this.context = context;
        deviceId = getDeviceIdInternal(context);
        deviceName = getDeviceNameInternal();
    }

    public AndroidDevice(Context context, String deviceId, String deviceName) {
        this.context = context;
        this.deviceId = deviceId;
        this.deviceName = deviceName;
    }

    @Override
    public String getDeviceName() {
        return deviceName;
    }

    @Override
    public String getDeviceId() {
        return deviceId;
    }

    private String getDeviceIdInternal(Context context) {
        return Settings.Secure.getString(context.getContentResolver(),
                Settings.Secure.ANDROID_ID);
    }

    private String getDeviceNameInternal() {
        String deviceName = Settings.Global.getString(context.getContentResolver(), "device_name");
        if (deviceName != null) {
            return deviceName;
        }

        String manufacturer = Build.MANUFACTURER;
        String model = Build.MODEL;
        if (model.startsWith(manufacturer)) {
            return model;
        } else {
            return manufacturer + " " + model;
        }
    }
}

File: android/src/main/java/org/jellyfin/apiclient/interaction/AndroidDevice.kt
package org.jellyfin.apiclient.interaction

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.provider.Settings
import org.jellyfin.apiclient.interaction.device.IDevice

data class AndroidDevice(
	override val deviceId: String,
	override val deviceName: String
) : IDevice {
	companion object {
		@SuppressLint("HardwareIds")
		fun getAutomaticId(context: Context): String =
			Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)

		fun getAutomaticName(context: Context): String {
			// Use name from device settings
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
				return Settings.Global.getString(context.contentResolver, Settings.Global.DEVICE_NAME)
			}

			// Concatenate the name based on manufacturer and model
			val manufacturer = Build.MANUFACTURER
			val model = Build.MODEL

			return if (model.startsWith(manufacturer)) model
			else "$manufacturer $model"
		}

		fun fromContext(context: Context) = AndroidDevice(
			deviceId = getAutomaticId(context),
			deviceName = getAutomaticName(context)
		)
	}
}
